[FIX] account: fix some translations related to chart template

svfu-odoo opened the pull request:
There are currently translation issues related to the chart template.
Due to this the names of some records do not receive the necessary
/ intended translations when installing a localization or new language.

When installing a new localization / chart some records have the following
problem with the values of their translatabe fields:
They are only installed in the language that was active when the localization / chart was installed.
Thus when switching languages (or using a different user with a different language)
the names are displayed in the "installation language".
Translations for all active languages should be installed (for all relevant records).

The same problem happens when installing a new language
(the same records do not receive a translation for the new language).

The translation issue concerns for example (some) accounts and journals;
see the (incomplete) list at the end of this message.

This commit tries to fix the translation issue for the translatable
fields of all relevant models.

Note!
=====
  * The translation mechanism only works for records with xmlid.
    If a module creates a record without xmlid it will not be translated.
  * The problem is only fixed for records with xmlid for which at least 1 of the
    following conditions holds:
        * The record (and the translatable field value) is defined in
          the body of the function decorated with 'https://github.com/template'
        * The translation of the value of the translatable field can
          be found in the module 'account' or in the module that
          is associated with the record (by 'ir.model.data')
    I.e. the problem is not solved for demo data: It is technically
    difficult to determine the module they originate from.
    This makes it difficult to load the right code translation (the
    module information is needed for this).
  * The translation mechanism is not necessarily triggered
    if the record is (in principal) part of the chart template
    but not installed as part of the chart template.
    This can i.e. happen if a module is installed after the
    localization / chart.
      * Example: account.journal "Salaries" from hr_payroll_account
  * We also want to "translate" / localize some untranslatable fields
    (like account journal codes). For these fields the terms will be
    installed in the language of the partner of the company for which
    the chart will be installed (fallback to lang / user lang from the
    env in case there is none set)
      * Currently there is no language set for many (all?) demo comany.
        Thus the values will remain in English for them (when the
        respective module is installed).

Examples / Details
==================

**Reproduce**:
  1. Switch to the French language (install if needed)
     (Settings App > General Settings > Languages)
  2. Install a localisation (e.g. l10n_fr).
  3. Check the French translations of the localization
       * Comptabilité > Configuration (Menu) > Journaux
         (Accounting > Configuration (Menu) > Journals)
           * Here the journal names are in French
       * Comptabilité > Configuration (Menu) > Plan comptable
         (Accounting > Configuration (Menu) > Chart of Accounts)
           * All the account names are in French
  4. Switch to English on the current user (or some other language)
     via the user profile on the top right.
  5. Check the names of the localization again
       * Accounting
           * The journal names are still in French
       * Accounting > Configuration (Menu) > Chart of Accounts
           * Some of the account names are still in French
           * E.g. "Compte d'attente de la banque" ("Bank Suspense Account")

Other things to test:
  * "Salaries" journal from enterprise module 'hr_payroll_account'
      * Not demo data; it will (partly) work after this commit (see "Note" above)
  * "IFRS Automatic transfers" journal from enterprise module
    'account_auto_transfer' (installed when installing l10n_fr)
      * Demo data; the problem remains after this commit

**Technically** the main problems are the following:
  1. The information of some of the created records is only defined in
     the code. Thus their translations have to be taken from the
     translation of the code.
     But at the point of translation it is not clear from which
     module the data came from. This is needed to load the right translation.
       * This was fixed for data from 'https://github.com/template' functions
  2. Some records are created without an xmlid and thus
     cannot be translated with the current translation mechanism at all.
       * This was fixed for the relevant records from module 'account'

Example Records
---------------

Some affected **accounts**:
  * from module 'account'
    * Bank utility accounts
      * Bank Suspense Account
      * Outstanding Receipts
      * Outstanding Payments
      * Cash Discount Loss
      * Cash Discount Gain
      * Cash Difference Loss
      * Cash Difference Gain
      * Liquidity Transfer
    * Bank / Cash journal default accounts
      * Bank
      * Cash
    * Unaffected earnings account
      * Undistributed Profits/Losses

Some affected **journals**
  * from module 'account'
    * Customer Invoices
    * Vendor Bills
    * Miscellaneuos Operations
    * Exchange Difference
    * Cash Basis Taxes
    * Bank
    * Cash
  * from module 'account_auto_transfer' (enterprise)
    * IFRS Automatic Transfers
      * The problem will remain since it is demo data
  * from module 'hr_payroll_account' (enterprise)
    * Salaries
      * The translation is only loaded if the module is installed
        before the localization / chart

task info
=========

task-3414329

malb-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -1002,3 +1020,27 @@ def _load_translations(self, langs=None, companies=None):
                                     translation_importer.model_translations[mname][fname][xml_id][lang] = value
         translation_importer.save(overwrite=False)
+
+    def _load_code_translations(self, accounts, langs=None):
```
Is there a reason we put langs as a parameter but don't use it ? i see that the function "_load_translations" and "_load_code_translations" in transifex_code_translation.py did the same.

Also is this function an overide of the one from transifex_code_translation ? in that case it may be missing a super ? or just rename the function a bit :smile:

svfu-odoo replied:
I renamed the function to `_load_account_name_code_translations` to avoid any ambiguity.

I used the `langs` parameter since it was used for `_load_translation`. But since it is not used anywhere I removed it.

malb-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -1002,3 +1020,27 @@ def _load_translations(self, langs=None, companies=None):
+            name_translations = {
+                lang_code: translations_dict[lang_code][account.name]
+                for lang_code in langs
+                if lang_code != 'en_US' and account.name in translations_dict[lang_code]
```
Maybe a comment here for the why the != en_US

malb-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -1002,3 +1020,27 @@ def _load_translations(self, langs=None, companies=None):
+        """
+
+        langs = langs or [code for code, _name in self.env['res.lang'].get_installed()]
+        translations_dict = {
```
Suggestion: Not really a fan of putting "dict" at the end of the variable because by reading the name of the variable we will supposed that it's a dict but for wathever reason the type change, it will be confusing

malb-odoo reviewed (changes requested):
Good job ! Small comments and it will be good :smile:

svfu-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -168,6 +168,10 @@ def _load(self, template_code, company, install_demo):
         )
         company = company.with_env(self.env)
 
+        old_accounts = self.env['account.account'].search([
```
Renamed to `preexisting_accounts`. With `existing_account` it seemed a bit weird in the computation of  `new_accounts_without_xmlid` since all accounts should exist at that point.

svfu-odoo commented:
Thanks for the feedback

malb-odoo reviewed (approved):
LGTM thx !

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -186,6 +188,18 @@ def _load(self, template_code, company, install_demo):
+        #     * (some) bank utility accounts (see function _setup_utility_bank_accounts)
+        #     * cash / bank journal accounts (created "by" account.journal)
+        #     * unaffected earnings accounts (created "by" res.company)
+        ids_of_accounts_with_xmlid = self.env['ir.model.data'].search([('model', '=', 'account.account')]).mapped('res_id')
+        new_accounts_without_xmlid = self.env['account.account'].search([
```
Why only for `account.account`? I didn't test, but I'm pretty sure that the journals have the same issue for instance.
Same for reconciliation models.

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -186,6 +188,18 @@ def _load(self, template_code, company, install_demo):
+            ('company_id', '=', company.id),
+            ('id', 'not in', preexisting_accounts.ids + ids_of_accounts_with_xmlid),
+        ])
+        self._load_account_name_code_translations(new_accounts_without_xmlid)
```
Don't we want to also translate those when we install a new language?
It would fit better in `<account.chart.template>._load_translations` IMO

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -186,6 +188,18 @@ def _load(self, template_code, company, install_demo):
+        #     * cash / bank journal accounts (created "by" account.journal)
+        #     * unaffected earnings accounts (created "by" res.company)
+        ids_of_accounts_with_xmlid = self.env['ir.model.data'].search([('model', '=', 'account.account')]).mapped('res_id')
+        new_accounts_without_xmlid = self.env['account.account'].search([
+            ('company_id', '=', company.id),
+            ('id', 'not in', preexisting_accounts.ids + ids_of_accounts_with_xmlid),
+        ])
```
Instead of searching for all those things, would it not be better to search for records that don't have `name-->{lang}`, and try to see which translations are available from there?

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -1002,3 +1016,26 @@ def _load_translations(self, langs=None, companies=None):
+                if lang_code != 'en_US' and account.name in translations[lang_code]
+            }
+            if name_translations:
+                account.update_field_translations('name', name_translations)
```
With my other suggestions, you can just feed the data from `get_python_translations` into the `TranslationImporter` of `_load_translations` and do that in batch

william-andre reviewed (commented):
@antoine162 You asked me about this yesterday, are you aware of this PR?

antoine162 commented:
> @antoine162 You asked me about this yesterday, are you aware of this PR?

oooh nice. @jco-odoo you'll be glad to know Sven is doing this :smile:

svfu-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -183,7 +191,6 @@ def _load(self, template_code, company, install_demo):
             install_demo = False
         data = self._pre_load_data(template_code, company, template_data, data)
         self._load_data(data)
-        self._load_translations(companies=company)
```
l moved it to the end of the function and did not find any problem with it. Is it a problem?

svfu-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -986,19 +1041,64 @@ def _load_translations(self, langs=None, companies=None):
+
+        # translate the name of TEMPLATE_MODELS records that are not created from the chart_template data
+
+        # We use all modules that modify the model of self ('account.chart.template') as translation source.
+        # The MRO (w/o duplicates) of the class / model is used to order the modules.
+        # ('dict' preservers the order of insertion of the keys in Python 3.7 +)
+        files_modifying_this_model = list(dict.fromkeys([
+            getfile(cls) for cls in self.__class__.__mro__
+            if getattr(cls, '_name', None) == self.__class__._name
+        ]))
+        modules_to_use_for_code_translations = dict.fromkeys([
+            (get_resource_from_path(path) or (None, ))[0]  # module to which 'path' belongs to or 'None'
+            for path in files_modifying_this_model
+        ])
+        modules_to_use_for_code_translations.pop(None, None)
+        modules_to_use_for_code_translations = list(modules_to_use_for_code_translations)
```
Since I do not know which module the name is defined in I just try all modules that modify the `account.chart.template`. This could lead to wrong translations or performance issues. Maybe is is too much or you have a better idea?

william-andre replied:
I think it is too much and we should only consider `account` for what is not defined with `@template`

svfu-odoo commented:
@william-andre

New different approach using only the `TranslationImporter` and trying to translate the name field of "all" models.
I added some questions above.

Main issues:
  * The code translations can only be queried per module. Which modules to query?
      * This is i.e. a poblem for records not from "chart template data" but my solution for `@template` data is also not 100% foolproof I think.
  * Records without xmlid are not translated currently
      * I added xmlids for the relevant records from the account module (only)
      * `@template` data should always have one I think.
      * Should I translate records without xmlid too?
        It would be more generic than having to add xmlids everywhere?
        But maybe it is "good enough" now

Please give me advise / suggestions on these issues.

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -560,12 +569,16 @@ def _get_chart_template_data(self, template_code):
                 key=lambda i: TEMPLATE_MODELS.index(i[0]) if i[0] in TEMPLATE_MODELS else 1000
             ):
                 for func in funcs:
-                    data = func(self, template_code)
+                    data = func(self.with_context(lang='en_US') if model in TEMPLATE_MODELS and self.env[model]._fields['name'].translate else self,
+                                template_code)
```
Not sure that we need to check `self.env[model]._fields['name'].translate`, the fact that it is in `TEMPLATE_MODEL` should be enough

```suggestion
                self_lang = self
                if model in TEMPLATE_MODELS and self.env[model]._fields['name'].translate:
                    self_lang = self.with_context(lang='en_US')
                for func in funcs:
                    data = func(self_lang, template_code)

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -972,6 +992,41 @@ def _parse_csv(self, template_code, model, module=None):
+            for lang in langs
+        ])
+
+        queries = [
+            f'''
+             SELECT '{model}' AS model,
+                    model_data.name AS xmlid,
+                    model.name AS name,
+                    model_data.module AS module
+             FROM {model_table} model
+             JOIN ir_model_data model_data
+                 ON  model_data.model = '{model}'
+                 AND model.id = model_data.res_id
+             WHERE {WHERE_clause_missing_translation}
+                 AND model.name->>'en_US' IS NOT NULL
+                 AND model.company_id IN ({','.join([str(id) for id in companies.ids])})
+            '''
+            for model_table, model in [(self.env[model]._table, model) for model in queried_models]
+        ]
```
```suggestion
        queries = [f"""
             SELECT '{model}' AS model,
                    model_data.name AS xmlid,
                    model.name AS name,
                    model_data.module AS module
               FROM {self.env[model]._table} model
               JOIN ir_model_data model_data ON model_data.model = '{model}'
                                            AND model.id = model_data.res_id
              WHERE {where_clause_missing_translation}
                AND model.name->>'en_US' IS NOT NULL
                AND model.company_id IN %(company_ids)s
        """ for model in queried_models]
```

william-andre replied:
It would also be better if `'{model}'` was properly escaped too, soon a lot easier in master

svfu-odoo replied:
Thanks for the update. I put `model` and `lang` (too) into the `params` in the new version.

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -972,6 +992,41 @@ def _parse_csv(self, template_code, model, module=None):
+            self.env[model].flush_model(['id', 'company_id', 'name'])
+        self.env['ir.model.data'].flush_model(['res_id', 'model', 'name'])
+
+        WHERE_clause_missing_translation = " OR ".join([
```
```suggestion
        where_clause_missing_translation = " OR ".join([
```

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -972,6 +992,41 @@ def _parse_csv(self, template_code, model, module=None):
+            for model_table, model in [(self.env[model]._table, model) for model in queried_models]
+        ]
+
+        self._cr.execute(' UNION ALL '.join(queries))
```
```suggestion
        self._cr.execute(' UNION ALL '.join(queries), {'company_ids': tuple(companies.ids)})
```

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -986,19 +1041,64 @@ def _load_translations(self, langs=None, companies=None):
                             value = record.get(f"{fname}@{lang}")
                             if not value:  # manage generic locale (i.e. `fr` instead of `fr_BE`)
                                 value = record.get(f"{fname}@{lang.split('_')[0]}")
+                            if not value and fname == 'name':
```
```suggestion
                            if not value:
```

svfu-odoo replied:
Sorry. Fixed

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -986,19 +1041,64 @@ def _load_translations(self, langs=None, companies=None):
             template_data.pop('template_data', None)
             for mname, data in template_data.items():
                 for _xml_id, record in data.items():
+                    _name_code_translation_module = record.pop('_name_code_translation_module', 'account')
```
```suggestion
                    translation_module = record.pop('__translation_module__', 'account')
```

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -986,19 +1041,64 @@ def _load_translations(self, langs=None, companies=None):
+                            if not value and fname == 'name':
+                                value = code_translations.get_python_translations(_name_code_translation_module, lang).get(record[fname])
+                                if not value:  # manage generic locale (i.e. `fr` instead of `fr_BE`)
+                                    value = code_translations.get_python_translations(_name_code_translation_module, lang.split('_')[0]).get(record[fname])
```
Define `generic_lang = lang.split('_')[0]` earlier since it is used twice now

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -986,19 +1041,64 @@ def _load_translations(self, langs=None, companies=None):
+
+        # translate the name of TEMPLATE_MODELS records that are not created from the chart_template data
+
+        # We use all modules that modify the model of self ('account.chart.template') as translation source.
```
It's not linked to `account.chart.template`, is it?

And for these values, isn't it always in `account`? Do you have an example for data that comes from another module but that is not declared with `@template`?

svfu-odoo replied:
It was mainly for the demo data.

I guess anything that should be part of the chart data could just be defined in an `@template` function anyway?

I only managed to find this 1 example that is directly linked to the chart template.
(But the names are not translated anyway currently and a bit technical maybe.)
https://github.com/odoo/enterprise/blob/fb85e33cb8853aa72ec3ac66f3e58d06c325fec2/l10n_ec_edi/models/account_chart_template.py

svfu-odoo commented on addons/account/models/company.py:
```diff
@@ -484,12 +485,18 @@ def get_unaffected_earnings_account(self):
-            })
+        return self.env['account.account']._load_records([
+            {
+                'xml_id': f"account.{str(self.id)}_unaffected_earnings_account_{uuid.uuid4().hex[:8]}",
```
I removed the UUID since there should only be one unaffected earnings account

svfu-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -47,6 +48,13 @@ def wrapper(*args, **kwargs):
+        path = func.__globals__['__file__']
+        path_info = get_resource_from_path(path)
+        module = path_info[0] if path_info else 'account'
+        api.attrsetter('_module', module)(wrapper)
```
I am not sure I understand this comment correctly.
If I do not call it the return value will be lost and nothing will happen?
I just made a chained call in the return value now.

svfu-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -560,12 +569,16 @@ def _get_chart_template_data(self, template_code):
                             template_data[model].update(data)
                         else:
                             for xmlid, record in data.items():
+                                if 'name' in record:
```
The same record (same xmlid) can be modified by multiple functions.
E.g. '_get_be_account_journal' from `l10n_be` modifies the 'sale' and 'purchase' journals but not their names.
These journals were previously defined and "named" in module `account`.
If we always update the translation module we end up with `l10n_be` which does not define the names of the 'sale' and 'purchase' journals (and thus does not have the translation).

I generalised the approach to all translatable fields now by making `__translation_module__` into a dctionary `field_name -> module`. But maybe it is too complicated?

Alternatively maybe just always taking the module of the function that "creates" the record would be good enough?

svfu-odoo commented:
New version
  * Fixed your issues (except see above)
  * Removed the computation of all the modules that modify account.chart.template
  * Generalised the code translation mechanism to all translatable fields (and not just 'name')

Should I also generalise the SQL query / second for loop to translate all translatable fields (and not just 'name')?

william-andre commented on addons/account/models/chart_template.py:
```diff
@@ -47,7 +49,13 @@ def wrapper(*args, **kwargs):
             return func(*args, **kwargs)
-        return api.attrsetter('_l10n_template', (template, model))(wrapper)
+
+        # the module the function originates from is used for code translations
+        path = func.__globals__['__file__']
+        path_info = get_resource_from_path(path)
+        module = path_info[0] if path_info else 'account'
+
+        return api.attrsetter('_l10n_template', (template, model))(api.attrsetter('_module', module)(wrapper))
```
I guess it's just to make it one line shorter, we kept the pattern of `odoo/api` but it might be more readable this way indeed :+1: 
My question was more about accessing `func.__globals__['__file__']`, do you see an issue there? It is used only to find the module with `get_python_translations`.

svfu-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -554,18 +562,33 @@ def _post_load_data(self, template_code, company, template_data):
-                    data = func(self, template_code)
+                    self_lang = self
+                    if translatable_fields:
+                        self_lang = self.with_context(lang='en_US')
```
I put the translation of the untranslatable fields at the end of `_pre_load_data`

svfu-odoo commented:
@william-andre Please have another look

william-andre commented on addons/account/tests/test_chart_template.py:
```diff
@@ -469,22 +470,173 @@ def get_domain(model):
+                    with patch.object(code_translations, 'python_translations', mock_python_translations):
+                        self.env['account.chart.template'].try_loading('translation', company=company, install_demo=False)
+
+        # Check translations
```
This is very complicated, which should be avoided as much as possible for tests. Business code can be complicated, but the test should be dead simple.
Can't we do something like this? (this is certainly too simple, but I'm sure you get the idea)

```python
fields_to_test = {'account.tax': ['name']}
self.assertEqual({
    f'{xmlid}.{field}@{lang}': self.env['account.chart.template'].ref(xmlid).with_context(lang=lang)[field]
    for lang in ['fr_BE', 'en_US']
    for model, data in chart_data.items():
    for xmlid in data
    for field in fields_to_test[model]
}, {
    'test_tax_1_template.name@en_US': "Tax 1",
    'test_tax_1_template.name@fr_BE': "Tax 1",
    ...
})
```

svfu-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -993,6 +1037,143 @@ def _parse_csv(self, template_code, model, module=None):
+        :rtype: dict[str, list[str]]
+        """
+        return {
+            'account.reconcile.model': [
+                'name',
+            ],
```
It is translatable starting with 17.0

svfu-odoo commented:
@william-andre  Please have another look

william-andre reviewed (approved):
@robodoo r+

Really good job :+1:

svfu-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -391,6 +408,21 @@ def _pre_load_data(self, template_code, company, template_data, data):
 
+        # Translate the untranslatable fields we want to translate anyway
+        untranslatable_model_fields = self._get_untranslatable_fields_to_translate()
+        untranslatable_target_lang = self._get_untranslatable_fields_target_language(template_code, company)
```
Thanks. I guess this is resolved since it is r+?

It would also be a possibility. There may not be a language in the context though (see comment in `_get_untranslatable_fields_target_language`).
Taking the company language (if possible) or using `get_lang` makes it more robust imo (it also was discussed at some point).
Having it as a function also makes it a bit more flexible.
